### PR TITLE
feat: add an `hasSize` function

### PR DESCRIPTION
**Pull request by alberic89**

### CODE REVIEW

File: src/root.zig [hunk 1]
```diff
@@ -210,6 +210,15 @@ pub fn AlgorithmType(
             self.max_needle = max_needle;
         }
 
+        // Check is there is enough memory allocated
```
Comment (fjebaker) [inline src/root.zig]:
```suggestion
        // Check if buffers have sufficient memory for a given haystack and 
        // needle length.
```

File: src/root.zig [hunk 2]
```diff
@@ -210,6 +210,15 @@ pub fn AlgorithmType(
...
+        pub fn hasSize(self: *Self, max_haystack: usize, max_needle: usize) bool {
+            if (self.max_haystack < max_haystack or self.max_needle < max_needle) {
+                return false;
+            } else {
+                return true;
+            }
```
Comment (fjebaker) [inline src/root.zig]:
```suggestion
            return (max_haystack <= self.max_haystack) and (max_needle <= self.max_needle);
```
It's easier just to check the positive case.

File: src/root.zig [hunk 3]
```diff
@@ -210,6 +210,15 @@ pub fn AlgorithmType(
             self.max_needle = max_needle;
         }
 
+        // Check is there is enough memory allocated
+        pub fn hasSize(self: *Self, max_haystack: usize, max_needle: usize) bool {
```
Comment (fjebaker) [inline src/root.zig]:
```suggestion
        pub fn hasSize(self: *const Self, max_haystack: usize, max_needle: usize) bool {
```
This won't and should not mutate memory, so we mark the pointer as `const`.

File: src/root.zig [hunk 4]
```diff
@@ -649,6 +658,11 @@ pub const Ascii = struct {
     pub fn resize(self: *Ascii, max_haystack: usize, max_needle: usize) !void {
         try self.alg.resize(max_haystack, max_needle);
     }
+
+    // Check is there is enough memory allocated
+    pub fn hasSize(self: *Ascii, max_haystack: usize, max_needle: usize) bool {
```
Comment (fjebaker) [inline src/root.zig]:
```suggestion
    // Check if buffers have sufficient memory for a given haystack and 
    // needle length.
    pub fn hasSize(self: *const Ascii, max_haystack: usize, max_needle: usize) bool {
```

File: src/unicode.zig [hunk 5]
```diff
@@ -219,6 +219,11 @@ pub const Unicode = struct {
     pub fn resize(self: *Unicode, max_haystack: usize, max_needle: usize) !void {
         try self.alg.resize(max_haystack, max_needle);
     }
+
+    // Check is there is enough memory allocated
+    pub fn hasSize(self: *Unicode, max_haystack: usize, max_needle: usize) bool {
```
Comment (fjebaker) [inline src/unicode.zig]:
```suggestion
    // Check if buffers have sufficient memory for a given haystack and 
    // needle length.
    pub fn hasSize(self: *const Unicode, max_haystack: usize, max_needle: usize) bool {
```

**Review by fjebaker — CHANGES_REQUESTED** [review]

**Comment by fjebaker** [comment]
In your previous PR you had some tests that made sure `resize` works. It would be very useful to include those (and similar for `hasSize`) in this PR too :+1:

**Comment by alberic89** [comment]
But just a question : is it an error that there is `<` and not `<=` at https://github.com/fjebaker/fuzzig/blob/2c5d606f679ee21bd42ef294f4d0be179f36d288/src/root.zig#L256 ?
It makes `score` fail when there is just the size needed.

**Comment by fjebaker** [comment]
> is it an error that there is < and not <= at 

Yes that is a mistake. Well spotted! Make sure there's a test that catches it and then I think this will be ready to merge.

Edit: I see you're ahead of me and already have!

File: src/root.zig [hunk 6]
```diff
@@ -849,3 +861,35 @@ test "traceback" {
...
+    try std.testing.expect(alg.hasSize(2, 2));
+    try alg.resize(2, 2);
+    try std.testing.expect(alg.hasSize(2, 2));
+    try doTestScore(&alg, "ab", "ab", o.score_match * 2 +
+        (o.bonus_head * o.bonus_first_character_multiplier) +
+        o.bonus_consecutive);
+
+    try std.testing.expect(!alg.hasSize(8, 2));
+    try alg.resize(8, 2);
+    try std.testing.expect(alg.hasSize(8, 2));
```
Comment (fjebaker) [inline src/root.zig:884]:
```suggestion
    try std.testing.expect(alg.hasSize(8, 2));
    // larger sizes should fail
    try std.testing.expect(!alg.hasSize(8, 3));
    // smaller sizes should be fine
    try std.testing.expect(alg.hasSize(2, 1));
```

**Review by fjebaker — APPROVED** [review]
This is all great! My only suggestion is to add a handful more tests that check that smaller sizes pass, and that larger sizes do indeed fail (even after resize).

When you're happy, let me know and we can merge.

**Comment by alberic89** [comment]
For me, all is good. We can merge it.

**Review by fjebaker — APPROVED** [review]

**Pull request merged**